plugins/batterystat.koplugin/main.lua

```diff
@@ -0,0 +1,205 @@
+    table.insert(kv_pairs, {_("    Total minutes"), shorten(self:minutes())})
+    table.insert(kv_pairs, {_("    % per hour"), shorten(self:percentagePerHour())})
+end
+
+local BatteryStat = WidgetContainer:new{
+    name = "batterstat",
+}
+
+function BatteryStat:init()
+    local records = G_reader_settings:readSetting("batterystat") or {}
+    self.charging = Usage:new(records.charging)
+    self.decharging = Usage:new(records.decharging)
+    self.awake = Usage:new(records.awake)
+    self.sleeping = Usage:new(records.sleeping)
+
+    -- Note: these fields are not the "real" timestamp and battery usage, but
+    -- the unaccumulated values.
+    self.charging_state = State:new(records.charging_state)
+    self.awake_state = State:new(records.awake_state)
+    -- Whether the device was suspending before current timestamp.
+    self.was_suspending = false
+    -- Whether the device was charging before current timestamp.
+    self.was_charging = PowerD:isCharging()
+
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function BatteryStat:onFlushSettings()
+    local records = {
+      self.charging,
+      self.decharging,
+      self.awake,
+      self.sleeping,
+      self.charging_state,
+      self.awake_state,
+    }
+    G_reader_settings:saveSetting("batterystat", records);
+end
+
+function BatteryStat:accumulate()
+    if self.was_suspending then
+        -- Suspending to awake.
+        self.sleeping:append(self.awake_state)
+    else
+        -- Awake to suspending, time between self.awake_state and now should belong to awake.
+        self.awake:append(self.awake_state)
+    end
+    if self.was_charging then
+        -- Decharging to charging.
+        self.charging:append(self.charging_state)
+    else
+        self.decharging:append(self.charging_state)
+    end
+    self.awake_state = State:new()
+    self.charging_state = State:new()
+end
+
+function BatteryStat:onSuspend()
+    self.was_suspending = false
+    self:accumulate()
+end
+
+function BatteryStat:onResume()
+    self.was_suspending = true
+    self:accumulate()
+end
+
+function BatteryStat:onStartCharging()
+    self.was_charging = false
+    self:dumpToText()
+    self.charging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:onStopCharging()
+    self.was_charging = true
+    self:dumpToText()
+    self.decharging = Usage:new()
+    self.awake = Usage:new()
+    self.sleeping = Usage:new()
+    self:accumulate()
+end
+
+function BatteryStat:dumpToText()
+    local kv_pairs = self:dump()
+    local content = T(_("Dump at %1"), os.date("%c", os.time()))
+    for _, pair in ipairs(kv_pairs) do
+        content = content .. "\n" .. pair[1]
+        if pair[2] ~= nil and pair[2] ~= "" then
+            content = content .. "\t" .. pair[2]
+        end
+    end
+    content = content .. "\n-=-=-=-=-=-\n"
+    local file = io.open(DataStorage:getDataDir() .. "/battery_stat.txt", "a")
```

Comment:
I agree with your calculation :) I was about to suggest only enable the batterystat logging when reader is running in debug mode, but the hour estimation feature sounds interesting. Is the code included in this PR?